app/waf/src/controllers/adc.controller.ts

```diff
@@ -720,74 +670,154 @@ export class AdcController extends BaseController {
       });
     }
   }
+
+  private async onboarding(adc: Adc, addon: AddonReqValues): Promise<void> {
+    try {
+      this.logger.debug('start to do onbarding');
+      await this.serialize(adc, {status: AdcState.ONBOARDING});
+
+      let doMgr = await OnboardingManager.instanlize(this.wafapp);
+      let doBody = await doMgr.assembleDo(adc, {onboarding: true});
+      let doId = await doMgr.onboarding(doBody);
+
+      await checkAndWait(() => doMgr.isDone(doId), 240)
+        .then(() =>
+          checkAndWait(() => this.adcStCtr.gotTo(AdcState.ONBOARDED), 240),
+        )
+        .then(() => this.serialize(adc, {status: AdcState.ONBOARDED}));
+    } catch (error) {
+      await this.serialize(adc, {
+        status: AdcState.ONBOARDERR,
+        lastErr: `${AdcState.ONBOARDERR}: ${error}`,
+      });
+    }
+  }
 }
 
-type AddonReqValues = {
+export type AddonReqValues = {
   userToken: string;
   tenantId: string;
 };
 
-class AdcStateCtrlr {
+export class AdcStateCtrlr {
   private states: AdcStateEntry[] = [
     {
       state: AdcState.NEW,
-      check: Promise.resolve(true),
+      check: (ctrl: AdcStateCtrlr) => {
+        return Promise.resolve(true);
+      },
       next: [AdcState.POWERON],
     },
     {
+      failure: AdcState.POWERERR,
       state: AdcState.POWERON,
-      check: this.accessible(),
+      check: this.accessible,
```

Comment:
I hope to know how can we implement error exit path in this structure. For instance, after creating a VE instance in OpenStack, if the instance is ACTIVE and not pingable, we need to wait and check again; if the instance become ERROR, we need to stop waiting and move ADC state to POWERERROR. How does the code do this?

Reply:
I am just curious if you already have an idea to support error exiting or not...

Reply:
We've got an initial idea on this, and @haoshen61  is working on this.

Reply:
If current code structure can not support it, let us do it in next pr.

Reply:
OK.